Add privacy and security considerations section (#252)
quicwg/qlog: draft-ietf-quic-qlog-main-schema.md

# Security and privacy considerations {#privacy}

Modern protocols such as QUIC ({{!RFC9000}}) take pains to encrypt a large part
of their wire image, leading to generally strong privacy and security
characteristics, especially considering monster-in-the-middle threat models. A
generalized, plaintext logging scheme such as qlog by its nature undoes some of
these characteristics to make (parts of) the protocols observable again. This is
needed for a variety of use cases such as (live) operational insight,
implementation debuggability or research, but carries inherent security and
privacy risks, especially if qlog data can be correlated with other data
sources. As such, qlog operators and implementers should be aware of these risks
inherent in logging, storing, using, and in some cases purposefully disclosing
potentially sensitive information as part of qlog data.

> A non-exhaustive list of potentially sensitive fields in qlog data include:

Reviewer: This is a good starter but it seems a little too informal and at the same too specific in citing QUIC. Also modern will soon be old :)

Here's an alternative that builds on what you've put down already

> Protocols such as TLS [cite] and QUIC [cite] provide secure protection, to varying degrees, for the wire image [cite  RFC 8546]. There is inevitably tension between security and observability, when logging can reveal aspects of the wire image, that would ordinarily be protected. This tension equally applies to any privacy considerations that build on security properties, especially if data can be correlated across data sources.

> qlog operators and implementers should be mindful of the security and privacy risks inherent in handling qlog data. This includes but is not limited to logging, storing, or using the data. Data might be considered as non-sensitive, potentially-sensitive, or sensitive; applying the considerations in this section may produce different risks depending on the nature of the data itself, or its handling. However, in many cases the largest risk factors arise from data that can be considered as potenially-sensitive or sensitive. The following is a non-exhaustive list of such fields that can be carried in qlog data:

* IP addresses and transport protocol port numbers, which can be used to
  identify individual connections, endpoints and potentially users.

* QUIC Connection IDs and contents of NEW_CONNECTION_ID frames, which can be
  used to identify and track users across geographical networks {{!RFC9000}}.

* Accurate event timestamps, event counts, packet and frame sizes, which can be
  used with other data sources (e.g., encrypted packet captures) to correlate
  qlog data to a specific connection or user.

* Token values, which can contain address validation details and previous

>   connection information which could allow correlation of multiple connections.

Author: As you mentioned elsewhere, I was trying to stay away from too much protocol (QUIC, TLS) specifics. This is meant to encompass the various QUIC tokens (e.g., retry, address validation) and also indeed session tickets. 

One solution could be to list a few as examples like:

```suggestion
  Token values (e.g., QUIC address validation and retry tokens, TLS session tickets), which can contain address validation details
```

* TLS (decryption) keys, which can be used with other data sources (e.g.,
  encrypted packet captures) to correlate qlog data to a specific connection or

>   user or leak additional information.

Author: I'm not sure I captured this 100% correctly. Probably worth a new look now.

* Full or partial, encrypted or plaintext raw packet and frame payloads (e.g.,
  HTTP Field values, HTTP response data, TLS SNI field values), which can

>   contain other sensitive information. In qlog, these are typically (but not

Author: Not sure what "this second sentence" is here exactly. Do you mean "HTTP Field values, HTTP response data, TLS SNI field values" or the `RawInfo` bit?

>   contain other sensitive information. In qlog, these are typically (but not

Author: I've now removed the RawInfo bit here and placed it at the very end. Still not sure that's what you were saying here though, so please double-check.

  exclusively) contained in fields of the RawInfo type (see {{raw-info}}). qlog
  users should be particularly hesitant to include this information in all but
  the most stringent conditions.

The simplest and most extreme form of protection against abuse of this
information is the complete deletion of a given field, which is equivalent to
not logging the field(s) in question. While deletion completely protects the
data in the deleted fields from the risk of compromise, it also reduces the
utility of the dataset as a whole. As such, a balance should be found between
logging these fields and the potential risks inherent in their (involuntary)
disclosure. This balance depends on the use case at hand (e.g., research
datasets have different requirements to live operational troubleshooting). As
such, operators SHOULD only capture the minimal subset of qlog data relevant to
their specific purpose. Consequently, qlog implementers SHOULD allow for
changing this subset in a dynamic fashion, ideally even on a per-connection
basis.

> For the remaining fields deemed necessary for the use case at hand, as per
> {{!RFC6973}}, operators must be aware that the data will be at risk of
> compromise. As such, measures should be taken to firstly reduce the risk of

Reviewer: ```suggestion
Any data that is determined to be necessary for a use case at hand could be 
logged or captured. As per
{{!RFC6973}}, operators must be aware that such data will be at risk of
compromise. As such, measures should be taken to firstly reduce the risk of
```

compromise and secondly reduce the risk of abuse of compromised data. While a
full discussion of both aspects is out of scope for this document, the following

> paragraphs discuss high level concepts that can be applied to qlog data.

Reviewer: ```suggestion
paragraphs discuss high-level conciderations that can be applied to qlog data.
```

> To reduce the risk of compromise, operators can take measures such as: limiting

Reviewer: maybe we should say "unauthorized access or compromise"?

Author: I think that it should be clear enough what we're using RFC6973 here, so I didn't change this. If that turns out to be a problem for others down the line we can revisit.

> the amount of time data is stored, encrypting data at rest, limiting access

Reviewer: ```suggestion
the length of time that data is stored, encrypting data in transit and at rest, limiting access
```

> rights to the data, and auditing data usage practices. qlog deployments SHOULD

Reviewer: SHOULD here also

have integrated options for easy qlog data deletion or (aggressive) aggregation.

> To reduce the risk of data abuse after compromise, sensitive fields can be

Reviewer: does it have to be just "sensitive fields"? Just saying data fields would likely hold true.

anonymized, pseudonymized, otherwise permutated/replaced, truncated,
(re-)encrypted, or aggregated. A partial discussion of applicable techniques
(especially for IP address information) can be found in Appendix B of

> {{!DNS-PRIVACY=RFC8932}}. Operators should however be aware that many of these

Reviewer: ```suggestion
{{!DNS-PRIVACY=RFC8932}}. Operators should, however, be aware that many of these
```

techniques have been shown to be insufficient to safeguard user privacy and/or

> to protect user identity, especially if qlog data would be compromised in large
> amounts or be correlated with other data sources. As such, techniques from the

Reviewer: ```suggestion
to protect user identity, especially if a qlog data set is large or easily correlated
against other data sources. Differential privacy techniques might provide provide
```

> field of differential privacy (TODO: add survey paper reference) should be
> applied to mitigate this risk to its maximal potential.

Reviewer: ```suggestion
mitigations to this risk; see (TODO: add survey paper reference).
```

Finally, in keeping with the trend of maximum end user agency, qlog operators
SHOULD take end user preferences into account. While active user participation
(as per {{!RFC6973}}) on a per-qlog basis is difficult, as logs are often capture
out-of-band to the main user interaction and intent, implementers should have a
mechanism for integrating the capture, storage and removal of qlogs with more
general, often existing, user consent and privacy control systems. This is
especially true when logging (plaintext) packet and frame payload, as this can

> contain highly sensitive information.

Reviewer: I find this one a little hard to align with as written. what is the trend? Will that trend be one in 10 years time?

It might be better to rephrase the start is in terms of qlog operators needing to consider something we didn't alread mention: then tention between their operational needs and end user rights or preferences (the rights is important here, such as cookie laws, GDPR, regional policies etc). We can then close out the paragraph information that might contain more information that is sensitive or confidential to the end user (where previously we talked abput sensitivity mainly through the lens of connections), and then use the packet or frame payloads as example of where that data might reside.

the length of time that data is stored, encrypting data in transit and at rest,
limiting access rights to the data, and auditing data usage practices. qlog
deployments that provide integrated options for automated or manual data
deletion and (aggressive) aggregation, improve the ability to minimize the risk
of compromise.

To reduce the risk of data abuse after compromise, data can be anonymized,
pseudonymized, otherwise permutated/replaced, truncated, (re-)encrypted, or
aggregated. A partial discussion of applicable techniques (especially for IP
address information) can be found in Appendix B of {{!DNS-PRIVACY=RFC8932}}.
Operators should, however, be aware that many of these techniques have been
shown to be insufficient to safeguard user privacy and/or to protect user
identity, especially if a qlog data set is large or easily correlated against
other data sources. Differential privacy techniques might provide mitigations to
this risk; see (TODO: add survey paper reference).

Finally, qlog operators should consider the interplay between their use case
needs and end user rights or preferences (e.g., in the form of cookie laws, the

> GDPR, and other regional policies). While active user participation (as

Reviewer 2: Do not mention laws.  Do not mention the GDPR.

indicated by {{!RFC6973}}) on a per-qlog basis is difficult, as logs are often
captured out-of-band to the main user interaction and intent, general user
expectations should be taken into account. qlog deployments that provide
mechanisms to integrate the capture, storage and removal of qlogs with more

> general, often pre-existing, user consent and privacy control systems, improve

Reviewer 2: This might be a personal bugbear, but I don't think that mention of "user consent" - particularly as it is applied here - is a good idea.  The idea that a typical user of a service might consent to this form of data collection is a little fanciful, especially given the nature of the risks.

Author: I wasn't sure if you had doubts about this entire paragraph of just the "user consent" terminology in this one sentence. 

Since I personally feel it's important to at least mention user preference and that qlog collection should be tied to that (as should all data collection about a user), I interpreted it as just the terminology and changed to "user preference" for now. I will open a separate issue to discuss this further.
